flashbax/vault/vault.py

```diff
@@ -0,0 +1,448 @@
+                ),
+            )
+
+            # Construct metadata
+            self._metadata = {
+                "version": VERSION,
+                "structure": serialised_experience_structure,
+                **(metadata_json_ready or {}),  # Allow user to save extra metadata
+            }
+            # Dump metadata to file
+            metadata_path.write_text(json.dumps(self._metadata))
+        else:
+            # If the vault does not exist already, and no experience_structure is provided to create
+            # a new vault, we cannot proceed.
+            raise ValueError(
+                "Vault does not exist and no experience_structure was provided."
+            )
+
+        # We must now build the tree structure from the metadata, whether the metadata was created
+        #  here or loaded from file
+        if experience_structure is None:
+            # Since the experience structure is not provided, we simply use the metadata as is.
+            #  The result will always be a dictionary.
+            self._tree_structure = self._metadata["structure"]
+        else:
+            # If experience structure is provided, we try deserialise into that structure
+            self._tree_structure = deserialize_tree(
+                self._metadata["structure"],
+                target=experience_structure,
+            )
+
+        # Each leaf of the fbx_state.experience maps to a data store, so we tree map over the
+        #  tree structure to create each of the data stores.
+        self._all_ds = jax.tree_util.tree_map_with_path(
+            lambda path, x: self._init_leaf(
+                name=_path_to_ds_name(path),
+                leaf=x,
+                create_ds=not base_path_exists,
+            ),
+            self._tree_structure,
+            # Tree structure uses a list [shape, dtype] as a leaf
+            is_leaf=lambda x: isinstance(x, list),
+        )
+
+        # We keep track of the last fbx buffer idx received
+        self._last_received_fbx_index = 0
+
+        # We store and load the vault index from a separate datastore
+        self._vault_index_ds = ts.open(
+            self._get_base_spec("vault_index"),
+            dtype=jnp.int32,
+            shape=(1,),
+            create=not base_path_exists,
+        ).result()
+        # Just read synchronously as it's one number
+        self.vault_index = int(self._vault_index_ds.read().result()[0])
+
+    def _get_base_spec(self, name: str) -> dict:
+        """Simple common specs for all datastores.
+
+        Args:
+            name (str): name of the datastore
+
+        Returns:
+            dict: config for the datastore
+        """
+        return {
+            "driver": "zarr",
+            "kvstore": {
+                "driver": "ocdbt",
+                "base": f"{DRIVER}{self._base_path}",
+                "path": name,
+            },
+        }
+
+    def _init_leaf(
+        self, name: str, leaf: list, create_ds: bool = False
```

Comment:
If you do change to the `ShapeDtypeStruct` then change this type, otherwise make it `List[str]`